File: app/src/main/java/com/theandroiddev/riffking/ThreadFragment.java
package com.theandroiddev.riffking;

import android.content.Context;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.youtube.player.YouTubeInitializationResult;
import com.google.android.youtube.player.YouTubePlayer;
import com.google.android.youtube.player.YouTubePlayerSupportFragment;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.mikhaellopez.circularimageview.CircularImageView;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import static com.theandroiddev.riffking.HomeFragment.KEY_LAYOUT_MANAGER;

public class ThreadFragment extends Fragment implements YouTubePlayer.OnInitializedListener {
    private static final String TAG = "ThreadFragment";

    private static final String ARG_PARAM1 = "threadId";
    static boolean liked;
    //TODO 29-09 15:20 CLEAN IT
    protected RecyclerView mRecyclerView;
    protected RecyclerView.LayoutManager mLayoutManager;
    protected CommentAdapter mCommentAdapter;
    protected List<Comment> comments;
    protected HomeFragment.LayoutManagerType mCurrentLayoutManagerType;
    protected Helper helper;
    YouTubePlayerSupportFragment mYoutubePlayerFragment;
    TextView threadTitleTv, threadContentTv, threadLikesNumberTv, threadViewsNumberTv, threadUserTv;
    CustomEditText threadCommentEt;
    ImageView threadLikeIv, threadCommentSendIv;
    CircularImageView threadUserIv;
    //DATABASE
    FirebaseDatabase database;
    DatabaseReference databaseReference;
    FirebaseAuth firebaseAuth;
    User user;
    FrameLayout youtubeFrameLayout;
    private String threadId;
    private OnFragmentInteractionListener mListener;
    private Thread thread;



    public ThreadFragment() {
    }

    public static ThreadFragment newInstance(String param1, String param2) {
        ThreadFragment fragment = new ThreadFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        fragment.setArguments(args);
        return fragment;
    }

    public static void expand(final View v) {
        v.measure(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT);
        final int targetHeight = v.getMeasuredHeight();

        // Older versions of android (pre API 21) cancel animations for views with a height of 0.
        v.getLayoutParams().height = 1;
        v.setVisibility(View.VISIBLE);
        Animation a = new Animation() {
            @Override
            protected void applyTransformation(float interpolatedTime, Transformation t) {
                v.getLayoutParams().height = interpolatedTime == 1
                        ? FrameLayout.LayoutParams.WRAP_CONTENT
                        : (int) (targetHeight * interpolatedTime);
                v.requestLayout();
            }

            @Override
            public boolean willChangeBounds() {
                return true;
            }
        };

        // 1dp/ms
        a.setDuration((int) (targetHeight / v.getContext().getResources().getDisplayMetrics().density));
        v.startAnimation(a);
    }

    public static void collapse(final View v) {
        final int initialHeight = v.getMeasuredHeight();

        Animation a = new Animation() {
            @Override
            protected void applyTransformation(float interpolatedTime, Transformation t) {
                if (interpolatedTime == 1) {
                    v.setVisibility(View.GONE);
                } else {
                    v.getLayoutParams().height = initialHeight - (int) (initialHeight * interpolatedTime);
                    v.requestLayout();
                }
            }

            @Override
            public boolean willChangeBounds() {
                return true;
            }
        };

        // 1dp/ms
        a.setDuration((int) (initialHeight / v.getContext().getResources().getDisplayMetrics().density));
        v.startAnimation(a);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        comments = new ArrayList<>();

        threadId = getArguments().getString(ARG_PARAM1);
        database = FirebaseDatabase.getInstance();
        firebaseAuth = FirebaseAuth.getInstance();
        databaseReference = database.getReference();

        databaseReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {

                thread = dataSnapshot.child("threads").child(threadId).getValue(Thread.class);
                if (thread != null) {
                    user = dataSnapshot.child("users").child(thread.getUserId()).getValue(User.class);
                    if (user != null) {
                        threadUserTv.setText(user.getName());
                        Picasso.with(getContext()).load(user.getPhotoUrl()).into(threadUserIv);

                        if (getContext() != null) {
                            if (dataSnapshot.child("threadLikes").child(getCurrentUserId()).child(getThreadId()).getValue(Boolean.class) != null) {
                                Log.d(TAG, "onDataChange: " + "already liked");
                                liked = true;
                                helper.setLiked(threadLikeIv);

                            } else {
                                Log.d(TAG, "onDataChange: not liked!");
                                liked = false;
                                helper.setUnliked(threadLikeIv);


                            }

                        }
                    }
                }

                //comments.clear();
                Iterable<DataSnapshot> children = dataSnapshot.child("comments").child(threadId).getChildren();

                for (DataSnapshot child : children) {
                    Comment comment = child.getValue(Comment.class);
                    comment.setId(child.getKey());
                    comments.add(0, comment);

                }
                Log.d(TAG, "onDataChange1: " + comments.toString());
                mCommentAdapter.notifyDataSetChanged();
                Log.e(TAG, "onDataChange:Comments refreshed! ");

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }

        });

        loadThread();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View fragmentThreadView = inflater.inflate(R.layout.fragment_thread, container, false);


        if (getArguments() != null) {

            mRecyclerView = (RecyclerView) fragmentThreadView.findViewById(R.id.thread_comments_rv);
            mLayoutManager = new LinearLayoutManager(getActivity());


            mCurrentLayoutManagerType = HomeFragment.LayoutManagerType.LINEAR_LAYOUT_MANAGER;
            if (savedInstanceState != null) {
                // Restore saved layout manager type.
                mCurrentLayoutManagerType = (HomeFragment.LayoutManagerType) savedInstanceState
                        .getSerializable(KEY_LAYOUT_MANAGER);
            }

            mCommentAdapter = new CommentAdapter(getContext(), comments, databaseReference, getCurrentUserId());
            Log.d(TAG, "onDataChange2: " + comments.toString());

            mRecyclerView.setNestedScrollingEnabled(false);
            mRecyclerView.setAdapter(mCommentAdapter);

            helper.setRecyclerViewLayoutManager(HomeFragment.LayoutManagerType.LINEAR_LAYOUT_MANAGER, mRecyclerView,
                    getActivity(), mLayoutManager, mCurrentLayoutManagerType);

            threadTitleTv = (TextView) fragmentThreadView.findViewById(R.id.thread_title_tv);
            threadLikesNumberTv = (TextView) fragmentThreadView.findViewById(R.id.thread_likes_number_tv);
            threadViewsNumberTv = (TextView) fragmentThreadView.findViewById(R.id.thread_views_number_tv);
            threadContentTv = (TextView) fragmentThreadView.findViewById(R.id.thread_description);
            threadLikeIv = (ImageView) fragmentThreadView.findViewById(R.id.thread_like_iv);
            threadCommentSendIv = (ImageView) fragmentThreadView.findViewById(R.id.thread_comment_send_iv);
            threadCommentEt = (CustomEditText) fragmentThreadView.findViewById(R.id.thread_comment_et);
            threadUserTv = (TextView) fragmentThreadView.findViewById(R.id.thread_user_tv);
            threadUserIv = (CircularImageView) fragmentThreadView.findViewById(R.id.thread_user_iv);
            youtubeFrameLayout = (FrameLayout) fragmentThreadView.findViewById(R.id.fragment_youtube_player);


            threadUserTv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openProfileFragment();
                }
            });

            threadUserIv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openProfileFragment();
                }
            });

            threadLikeIv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    if (!thread.getUserId().equals(getCurrentUserId())) {
                    if (!liked) {
                        helper.transaction(databaseReference.child("threads").child(threadId).child("likes"), 1);
                        helper.transaction(databaseReference.child("users").child(thread.getUserId()).child("likes"), 1);
                        databaseReference.child("threadLikes").child(getCurrentUserId()).child(threadId).setValue(true);

                    } else {
                        helper.transaction(databaseReference.child("threads").child(threadId).child("likes"), -1);
                        helper.transaction(databaseReference.child("users").child(thread.getUserId()).child("likes"), -1);
                        databaseReference.child("threadLikes").child(getCurrentUserId()).child(threadId).removeValue();
                    }

                }
                }
            });

            threadCommentSendIv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (!TextUtils.isEmpty(threadCommentEt.getText().toString())) {
                        Comment comment = new Comment(threadId, getCurrentUserId(), threadCommentEt.getText().toString(), helper.getCurrentDate(), 0);
                        databaseReference.child("comments").child(getThreadId()).push().setValue(comment);
                        threadCommentEt.setText("");

                    }
                }
            });


            mYoutubePlayerFragment = new YouTubePlayerSupportFragment();
            mYoutubePlayerFragment.initialize(PlayerConfig.API_KEY, this);
            FragmentManager fragmentManager = getFragmentManager();
            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
            fragmentTransaction.replace(R.id.fragment_youtube_player, mYoutubePlayerFragment);
            fragmentTransaction.commit();

            helper.transaction(databaseReference.child("threads").child(threadId).child("views"), 1);

            return fragmentThreadView;

        }

        return inflater.inflate(com.theandroiddev.riffking.R.layout.fragment_thread, container, false);
    }

    private void openProfileFragment() {

        ProfileFragment profileFragment = new ProfileFragment();
        Bundle bundle = new Bundle();
        bundle.putString("USER_ID", thread.getUserId());
        bundle.putString("CURRENT_USER_ID", getCurrentUserId()); //Two same only in this case
        profileFragment.setArguments(bundle);

        FragmentTransaction fragmentTransaction = ((HomeActivity) getContext()).getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.content_home, profileFragment).addToBackStack(null);
        fragmentTransaction.commit();
    }

    private void loadThread() {

        databaseReference.child("threads").child(threadId).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {

                thread = dataSnapshot.getValue(Thread.class);

                if (thread == null) {
                    threadTitleTv.setText("Something went wrong...");
                } else {
                    threadTitleTv.setText(thread.getTitle());
                    setContent();
                    threadLikesNumberTv.setText(String.valueOf(thread.getLikes()));
                    threadViewsNumberTv.setText(String.valueOf(thread.getViews()));
                }

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });

    }

    private void setContent() {

        threadContentTv.setText(thread.getDescription());
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onPause() {

        mYoutubePlayerFragment.onPause();

        //FragmentManager fm = getFragmentManager();

        //fm.beginTransaction().remove(mYoutubePlayerFragment).commit();

        databaseReference.child("threads").child(threadId).setValue(thread);


        super.onPause();
    }


    @Override
    public void onResume() {
        super.onResume();

        HomeActivity homeActivity = (HomeActivity) getActivity();
        homeActivity.fab.setVisibility(View.GONE);



    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
            helper = new Helper(context);
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }

    }

    @Override
    public void onDetach() {
        //mYoutubePlayerFragment.onDetach();
        super.onDetach();
        mListener = null;
    }


    @Override
    public void onInitializationSuccess(YouTubePlayer.Provider provider, YouTubePlayer youTubePlayer, boolean b) {
        if (!b) {
            if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
                youTubePlayer.setFullscreen(true);
            } else youTubePlayer.setFullscreen(false);

            //cue instead of load to stop auto-play

            youTubePlayer.loadVideo(thread.getYoutubeId());
            youTubePlayer.setPlayerStyle(YouTubePlayer.PlayerStyle.DEFAULT);

        }
    }


    @Override
    public void onInitializationFailure(YouTubePlayer.Provider provider, YouTubeInitializationResult youTubeInitializationResult) {

        if (youTubeInitializationResult.isUserRecoverableError()) {
            youTubeInitializationResult.getErrorDialog(this.getActivity(), 1).show();
        } else {
            Toast.makeText(this.getActivity(),
                    "YouTubePlayer.onInitializationFailure(): " + youTubeInitializationResult.toString(),
                    Toast.LENGTH_LONG).show();
        }

    }


    public String getCurrentUserId() {
        Log.d(TAG, "TESTTTT: " + firebaseAuth.getCurrentUser().getUid());
        return firebaseAuth.getCurrentUser().getUid();
    }

    public String getThreadId() {
        //TODO MAKE STH WITH THIS THREADID VAR
        return threadId;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/theandroiddev/riffking/ThreadFragment.kt
package com.theandroiddev.riffking

import android.content.Context
import android.content.res.Configuration
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.Transformation
import android.widget.FrameLayout
import android.widget.Toast
import com.google.android.youtube.player.YouTubeInitializationResult
import com.google.android.youtube.player.YouTubePlayer
import com.google.android.youtube.player.YouTubePlayerSupportFragment
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.squareup.picasso.Picasso
import com.theandroiddev.riffking.HomeFragment.Companion.KEY_LAYOUT_MANAGER
import kotlinx.android.synthetic.main.fragment_home.*
import kotlinx.android.synthetic.main.fragment_thread.*
import java.util.*


class ThreadFragment : Fragment(), YouTubePlayer.OnInitializedListener {
    //TODO 29-09 15:20 CLEAN IT
    protected var layoutManager: RecyclerView.LayoutManager? = null
    protected var commentAdapter: CommentAdapter? = null
    protected var comments: MutableList<Comment> = mutableListOf()
    protected var currentLayoutManagerType: HomeFragment.LayoutManagerType? = null
    protected var helper: Helper? = null
    internal var youtubePlayerFragment: YouTubePlayerSupportFragment? = null
    //DATABASE
    internal var database: FirebaseDatabase? = null
    internal var databaseReference: DatabaseReference? = null
    internal var firebaseAuth: FirebaseAuth? = null
    internal var user: User? = null
    //TODO MAKE STH WITH THIS THREADID VAR
    var threadId: String? = null

    private var listener: OnFragmentInteractionListener? = null
    private var thread: Thread? = null

    val currentUserId: String?
        get() {
            Log.d(TAG, "TESTTTT: " + firebaseAuth?.currentUser?.uid)
            return firebaseAuth?.currentUser?.uid
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        comments = ArrayList()

        threadId = arguments?.getString(ARG_PARAM1)
        database = FirebaseDatabase.getInstance()
        firebaseAuth = FirebaseAuth.getInstance()
        databaseReference = database?.reference

        databaseReference?.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val threadId = threadId
                val threadUserId = thread?.userId
                val currentUserId = currentUserId

                if (threadId != null && currentUserId != null && threadUserId != null) {
                    thread = dataSnapshot.child("threads").child(threadId).getValue(Thread::class.java)
                    if (thread != null) {
                        user = dataSnapshot.child("users").child(threadUserId).getValue(User::class.java)
                        if (user != null) {
                            thread_user_tv.text = user?.name
                            Picasso.get().load(user?.photoUrl).into(thread_user_iv)

                            if (context != null) {
                                if (dataSnapshot.child("threadLikes").child(currentUserId).child(threadId)
                                                .getValue(Boolean::class.java) != null) {
                                    Log.d(TAG, "onDataChange: " + "already liked")
                                    liked = true
                                    helper?.setLiked(thread_like_iv)

                                } else {
                                    Log.d(TAG, "onDataChange: not liked!")
                                    liked = false
                                    helper?.setUnliked(thread_like_iv)

                                }

                            }
                        }
                    }

                    comments.clear()
                    val children = dataSnapshot.child("comments").child(threadId).children

                    for (child in children) {
                        val comment = child.getValue(Comment::class.java)
                        comment?.id = child.key
                        if (comment != null) {
                            comments.add(0, comment)
                        }
                    }
                    Log.d(TAG, "onDataChange1: " + comments.toString())
                    commentAdapter?.notifyDataSetChanged()
                    Log.e(TAG, "onDataChange:Comments refreshed! ")

                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })

        loadThread()

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val fragmentThreadView = inflater.inflate(R.layout.fragment_thread, container, false)

        if (arguments != null) {

            layoutManager = LinearLayoutManager(activity)

            currentLayoutManagerType = HomeFragment.LayoutManagerType.LINEAR_LAYOUT_MANAGER
            if (savedInstanceState != null) {
                // Restore saved layout manager type.
                currentLayoutManagerType = savedInstanceState
                        .getSerializable(KEY_LAYOUT_MANAGER) as HomeFragment.LayoutManagerType
            }

            val context = context
            val activity = activity
            val layoutManager = layoutManager
            val currentLayoutManagerType = currentLayoutManagerType
            val databaseReference = databaseReference
            val currentUserId = currentUserId

            if (context != null && activity != null && layoutManager != null && currentUserId != null
                    && currentLayoutManagerType != null && databaseReference != null) {
                commentAdapter = CommentAdapter(context, comments, databaseReference, currentUserId)
                Log.d(TAG, "onDataChange2: " + comments.toString())

                recyclerView.isNestedScrollingEnabled = false
                recyclerView.adapter = commentAdapter

                helper?.setRecyclerViewLayoutManager(HomeFragment.LayoutManagerType.LINEAR_LAYOUT_MANAGER, recyclerView,
                        activity, layoutManager, currentLayoutManagerType)
            }
            thread_user_tv.setOnClickListener { openProfileFragment() }

            thread_user_iv.setOnClickListener { openProfileFragment() }

            thread_like_iv.setOnClickListener {

                val threadId = threadId
                val threadUserId = thread?.userId

                if (threadUserId != currentUserId && threadUserId != null && threadId != null) {
                    val threadLikes = databaseReference?.child("threads")?.child(threadId)?.child("likes")
                    val userLikes = databaseReference?.child("users")?.child(threadUserId)?.child("likes")
                    if (threadLikes != null && userLikes != null && currentUserId != null) {

                        if (!liked) {

                            helper?.transaction(threadLikes, 1)
                            helper?.transaction(userLikes, 1)
                            databaseReference.child("threadLikes").child(currentUserId).child(threadId).setValue(true)

                        } else {
                            helper?.transaction(threadLikes, -1)
                            helper?.transaction(userLikes, -1)
                            databaseReference.child("threadLikes").child(currentUserId).child(threadId).removeValue()
                        }
                    }

                }
            }

            thread_comment_send_iv.setOnClickListener {
                if (!TextUtils.isEmpty(thread_comment_et.text.toString())) {
                    val threadId = threadId
                    if (threadId != null && currentUserId != null) {
                        val comment = Comment(threadId, currentUserId, thread_comment_et.text.toString(),
                                helper?.currentDate ?: "N/A", 0)
                        databaseReference?.child("comments")?.child(threadId)?.push()?.setValue(comment)
                        thread_comment_et.setText("")
                    }

                }
            }


            youtubePlayerFragment = YouTubePlayerSupportFragment()
            youtubePlayerFragment?.initialize(PlayerConfig.API_KEY, this)
            val fragmentManager = fragmentManager
            val fragmentTransaction = fragmentManager?.beginTransaction()

            val youtubePlayerFragment = youtubePlayerFragment
            val threadId = threadId
            if (youtubePlayerFragment != null && threadId != null) {
                fragmentTransaction?.replace(R.id.fragment_youtube_player, youtubePlayerFragment)
                fragmentTransaction?.commit()

                val threadViews = databaseReference?.child("threads")?.child(threadId)?.child("views")
                if (threadViews != null) {
                    helper?.transaction(threadViews, 1)
                } else {
                    //TODO error
                }
            }
            return fragmentThreadView

        }

        return inflater.inflate(com.theandroiddev.riffking.R.layout.fragment_thread, container, false)
    }

    private fun openProfileFragment() {

        val profileFragment = ProfileFragment()
        val bundle = Bundle()
        bundle.putString("USER_ID", thread?.userId)
        bundle.putString("CURRENT_USER_ID", currentUserId) //Two same only in this case
        profileFragment.arguments = bundle

        val fragmentTransaction = (context as HomeActivity).supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.content_home, profileFragment).addToBackStack(null)
        fragmentTransaction.commit()
    }

    private fun loadThread() {

        val threadId = threadId
        if (threadId != null) {
            databaseReference?.child("threads")?.child(threadId)?.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {

                    thread = dataSnapshot.getValue(Thread::class.java)

                    if (thread == null) {
                        thread_title_tv.text = "Something went wrong..."
                    } else {
                        thread_title_tv.text = thread?.title
                        setContent()
                        thread_likes_number_tv.text = thread?.likes.toString()
                        thread_views_number_tv.text = thread?.views.toString()
                    }

                }

                override fun onCancelled(databaseError: DatabaseError) {

                }
            })
        }

    }

    private fun setContent() {

        thread_description.text = thread?.description
    }

    // TODO: Rename method, update argument and hook method into UI event
    fun onButtonPressed(uri: Uri) {
        if (listener != null) {
            listener?.onFragmentInteraction(uri)
        }
    }

    override fun onPause() {

        youtubePlayerFragment?.onPause()

        //FragmentManager fm = getFragmentManager();

        //fm.beginTransaction().remove(youtubePlayerFragment).commit();
        val threadId = threadId
        if (threadId != null) {
            databaseReference?.child("threads")?.child(threadId)?.setValue(thread)
        }

        super.onPause()
    }


    override fun onResume() {
        super.onResume()

        val homeActivity = activity as HomeActivity?
        homeActivity?.fab?.visibility = View.GONE


    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            listener = context
            helper = Helper(context)
        } else {
            throw RuntimeException(context.toString() + " must implement OnFragmentInteractionListener")
        }

    }

    override fun onDetach() {
        //youtubePlayerFragment.onDetach();
        super.onDetach()
        listener = null
    }


    override fun onInitializationSuccess(provider: YouTubePlayer.Provider, youTubePlayer: YouTubePlayer, b: Boolean) {
        if (!b) {
            if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
                youTubePlayer.setFullscreen(true)
            } else
                youTubePlayer.setFullscreen(false)

            //cue instead of load to stop auto-play

            youTubePlayer.loadVideo(thread?.youtubeId)
            youTubePlayer.setPlayerStyle(YouTubePlayer.PlayerStyle.DEFAULT)

        }
    }


    override fun onInitializationFailure(provider: YouTubePlayer.Provider,
                                         youTubeInitializationResult: YouTubeInitializationResult) {

        if (youTubeInitializationResult.isUserRecoverableError) {
            youTubeInitializationResult.getErrorDialog(this.activity, 1).show()
        } else {
            Toast.makeText(this.activity,
                    "YouTubePlayer.onInitializationFailure(): " + youTubeInitializationResult.toString(),
                    Toast.LENGTH_LONG).show()
        }

    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        fun onFragmentInteraction(uri: Uri)
    }

    companion object {
        private val TAG = "ThreadFragment"

        private val ARG_PARAM1 = "threadId"
        internal var liked: Boolean = false

        fun newInstance(param1: String, param2: String): ThreadFragment {
            val fragment = ThreadFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, param1)
            fragment.arguments = args
            return fragment
        }

        fun expand(v: View) {
            v.measure(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
            val targetHeight = v.measuredHeight

            // Older versions of android (pre API 21) cancel animations for views with a height of 0.
            v.layoutParams.height = 1
            v.visibility = View.VISIBLE
            val a = object : Animation() {
                override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
                    v.layoutParams.height = if (interpolatedTime == 1f)
                        FrameLayout.LayoutParams.WRAP_CONTENT
                    else
                        (targetHeight * interpolatedTime).toInt()
                    v.requestLayout()
                }

                override fun willChangeBounds(): Boolean {
                    return true
                }
            }

            // 1dp/ms
            a.duration = (targetHeight / v.context.resources.displayMetrics.density).toInt().toLong()
            v.startAnimation(a)
        }

        fun collapse(v: View) {
            val initialHeight = v.measuredHeight

            val a = object : Animation() {
                override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
                    if (interpolatedTime == 1f) {
                        v.visibility = View.GONE
                    } else {
                        v.layoutParams.height = initialHeight - (initialHeight * interpolatedTime).toInt()
                        v.requestLayout()
                    }
                }

                override fun willChangeBounds(): Boolean {
                    return true
                }
            }

            // 1dp/ms
            a.duration = (initialHeight / v.context.resources.displayMetrics.density).toInt().toLong()
            v.startAnimation(a)
        }
    }
}
